tools/codegen/gen.py

```diff
@@ -347,73 +324,43 @@ def go(f: NativeFunction) -> Optional[str]:
         if Variant.function not in f.variants:
             return None
 
-        cpp_returns_type = cpp.returns_type(f.func.returns)
-        cpp_name = cpp.name(f.func)
-        signature_group = cpp.signature_group(f.func, method=False)
+        name = cpp.name(f.func)
+
+        sig_group = CppSignatureGroup.from_schema(f.func, method=False)
 
         if target is Target.DECLARATION:
-            if signature_group.gathered_signature is None:
-                # There's no TensorOptions
-                return f"""
-CAFFE2_API {cpp_returns_type} {cpp_name}({signature_group.signature.cpp_arguments_str(with_defaults=True)});
-"""
-            else:
-                # There's TensorOptions in the API. Create 2 APIs - one taking the TensorOptions object ("gathered_signature"),
-                # and one taking a scattered signature with ScalarType, Layout, Device separately ("signature").
-                # The gathered_signature already exists in several older PyTorch versions and had default arguments.
-                # For backward compatibility, we left it unchanged and added the scattered API on top of it.
-                # Note that the scattered API cannot have default arguments or calls will be ambigious.
-                return f"""
-CAFFE2_API {cpp_returns_type} {cpp_name}({signature_group.gathered_signature.cpp_arguments_str(with_defaults=True)});
-CAFFE2_API {cpp_returns_type} {cpp_name}({signature_group.signature.cpp_arguments_str(with_defaults=False)});
-"""
+            result = f"\nCAFFE2_API {sig_group.signature.decl()};\n"
```

Comment:
Does this work? Looks like you removed the defaults handling - if you don't have a faithful signature (i.e. faithful and non-faithful are identical), then the declaration must have defaults. If faithful and non-faithful differ, then the non-faithful must have defaults and the faithful must not have defaults. Breaking those rules breaks bc.

Reply:
This works because I moved the defaults handling into the construction of the signature groups. 

```
        if faithful:
            # Faithful signatures will ungroup arguments into argument
            # packs.
            #
            # After this, manually do overload disambiguation, by
            # dropping defaults from the faithful signature.  In
            # principle, we should be able to do this at some later
            # point in time with other overload disambiguation
            argument_packs = tuple(
                cpp.argument_faithful(a).no_default() for a in arguments
            )   
        else:   
            argument_packs = tuple(
                cpp.argument(a) for a in arguments
            )  
```

I think handling it at construction time is better, because it preserves the invariant "a CppSignature knows exactly how it should be rendered in C++"